test/performance/range/gap_decorator_benchmark.cpp

```diff
@@ -0,0 +1,452 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+#include <chrono>
+#include <cmath>
+#include <cstring>
+#include <random>
+#include <utility>
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/alignment/exception.hpp>
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/range/decorator/all.hpp>
+#include <seqan3/std/ranges>
+
+using namespace seqan3;
+
+typedef long double time_type;
+
+/* Helper function to sample gap length for each ungapped sequence position.
+ *
+ * Parameter:
+ * size_type    size type of underlying sequence range
+ *
+ * Arguments:
+ * gap_vector   reference to empty vector storing the sampled gap lengths
+ * size         final (aligned) sequence size
+ */
+template<typename size_type>
+void sample(std::vector<size_type> & gap_vector, size_type size)
+{
+    std::default_random_engine generator;
+    generator.seed(std::chrono::system_clock::now().time_since_epoch().count());
+    std::uniform_real_distribution<double> uni(0.0, 1.0);
+    std::array<double,10> cumsum = {0.6395, 0.8263, 0.8871, 0.9257, 0.9544, 0.9709, 0.9813, 0.9890, 0.9955, 1.0000};
+    for (size_type i = 0; i < size; ++i)
+    {
+        double y = uni(generator);
+        gap_vector[i] = y;
+        auto it = std::find_if(cumsum.begin(), cumsum.end(), [y](double i){return y <= i;});
```

Comment:
Line 45 is obsolete. Otherwise, this is how you sample from an arbitrary cumulative distribution function (cdf) with a uniform number generator, i.e. you have the cdf of the probability density function (from the paper), which is basically a step function and you sample now uniformly distributed the y-value in the range [0;1]. You then determine which step is "hit" by the y-value and that's the first one which is greater or equal to your uniformly sampled double.  I rename the lambda-local `i` for improving readability.